### ISSUE TITLE
crash with free(): invalid next size (normal): 0x00002aaad8002310

### ISSUE THREAD

**Issue by glennklockwood** [breuner/elbencho#30]
I seem to be reproducibly crashing elbencho and getting the following error:

```
WRITE: 0 files/s; 65 MiB/s; 0 files; 2978 MiB; 16 threads; 1% CPU; 46s*** Error in `/global/u2/g/glock/src/elbencho/bin.cgpu/elbencho': free(): invalid next size (normal): 0x00002aaad8002310 ***
FAULT HANDLER (PID 66952 / TID 66986): Abnormal termination
terminate called after throwing an instance of 'ProgException'
  what():  Backtrace:
1: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN8SignalTk12logBacktraceB5cxx11Ev+0x169) [0x4cb579]
2: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN8SignalTk18faultSignalHandlerEi+0xa8) [0x4cc158]
3: /lib64/libc.so.6(+0x361e0) [0x2aaaac8f71e0]
4: /lib64/libc.so.6(gsignal+0x110) [0x2aaaac8f7160]
5: /lib64/libc.so.6(abort+0x151) [0x2aaaac8f8741]
6: /lib64/libc.so.6(+0x794a7) [0x2aaaac93a4a7]
7: /lib64/libc.so.6(+0x7fcc3) [0x2aaaac940cc3]
8: /lib64/libc.so.6(+0x815c9) [0x2aaaac9425c9]
9: /usr/lib64/libstdc++.so.6(_ZNSt13basic_filebufIcSt11char_traitsIcEE26_M_destroy_internal_bufferEv+0x1e) [0x2aaaabebc64e]
10: /usr/lib64/libstdc++.so.6(_ZNSt13basic_filebufIcSt11char_traitsIcEE5closeEv+0x4b) [0x2aaaabebf9eb]
11: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN7CPUUtil6updateEv+0x262) [0x525ff2]
12: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN17WorkersSharedData25incNumWorkersDoneUnlockedEv+0x5c) [0x4e17cc]
13: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN6Worker17incNumWorkersDoneEv+0x4b) [0x4f478b]
14: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN11LocalWorker3runEv+0x12e0) [0x4f26e0]
15: /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho(_ZN6Worker11threadStartEPS_+0x12) [0x4f43f2]
16: /usr/lib64/libstdc++.so.6(+0xc338f) [0x2aaaabe8c38f]
17: /lib64/libpthread.so.0(+0x7569) [0x2aaaac6aa569]
18: /lib64/libc.so.6(clone+0x3f) [0x2aaaac9b9a2f]

Error: HTTP client error on benchmark interruption: Connection refused. Service: cgpu08:1611
HTTP client error in benchmark phase: Connection refused. Service: cgpu08:1611
srun: error: cgpu08: task 0: Aborted
srun: launch/slurm: _step_signal: Terminating StepId=2794884.0
```

This appears to happen once my job reaches the `--timelimit` under specific circumstances.  I run my job using the following:

```
srun -N 1 -n 1 --nodelist cgpu08 /global/u2/g/glock/src/elbencho/bin.cgpu/elbencho --service --foreground &

/global/u2/g/glock/src/elbencho/bin.cgpu/elbencho --hosts cgpu08 --timelimit 45 --blockvaralgo fast --blockvarpct 25 --resfile xyz --csvfile xyz.out.csv --threads 16 --files 1 --dirsharing --mkdirs --sync --direct --write --size 1t --block 4k --rand /vast/glock
```

This is using elbencho from commit c997820d263c229f8932d8f0789489a3233c6bf0 using gcc (SUSE Linux) 7.5.0 and boost-1.74.0.

I *think* I've been sporadically hitting the same error under different levels of concurrency but for some reason the stack trace doesn't make it out to the stderr of my batch job and I had to run elbencho interactively to get the above error.  For example, the following configuration failed with a similar "HTTP client error on benchmark interruption" error:

```
/global/u2/g/glock/src/elbencho/bin.cgpu/elbencho --hosts cgpu05,cgpu06,cgpu07,cgpu12,cgpu13,cgpu14 --timelimit 45 --resfile secondary_quiet.2p-6s.2780749.out.results --csvfile secondary_quiet.2p-6s.2780749.out.csv --livecsv secondary_quiet.2p-6s.2780749.out.live.csv --threads 16 --files 1 --dirsharing --mkdirs --sync --direct --write --size 1t --block 4k --rand /vast/glock
```

**Comment by glennklockwood** [comment]
I reproduced this with the unstripped binary through gdb and got the following stack trace:

```
(gdb) bt
#0  0x00002aaaac94d420 in raise () from /lib64/libc.so.6
#1  0x00002aaaac94ea01 in abort () from /lib64/libc.so.6
#2  0x00002aaaac990877 in __libc_message () from /lib64/libc.so.6
#3  0x00002aaaac997093 in malloc_printerr () from /lib64/libc.so.6
#4  0x00002aaaac998999 in _int_free () from /lib64/libc.so.6
#5  0x00002aaaabed3912 in std::basic_filebuf<char, std::char_traits<char> >::_M_destroy_internal_buffer() () from /usr/lib64/libstdc++.so.6
#6  0x00002aaaabed6ddb in std::basic_filebuf<char, std::char_traits<char> >::close() () from /usr/lib64/libstdc++.so.6
#7  0x0000000000525ff2 in std::basic_filebuf<char, std::char_traits<char> >::~basic_filebuf (this=0x2aaaadcd8a30, __in_chrg=<optimized out>) at /usr/include/c++/7/fstream:239
#8  std::basic_ifstream<char, std::char_traits<char> >::~basic_ifstream (this=0x2aaaadcd8a20, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /usr/include/c++/7/fstream:534
#9  CPUUtil::update (this=this@entry=0x7fffffff5e90) at source/CPUUtil.cpp:38
#10 0x00000000004e17cc in WorkersSharedData::incNumWorkersDoneUnlocked (this=0x7fffffff5df0) at source/workers/WorkersSharedData.cpp:17
#11 0x00000000004f478b in Worker::incNumWorkersDone (this=0x7b5920) at source/workers/Worker.cpp:27
#12 0x00000000004f26e0 in LocalWorker::run (this=0x7b5920) at source/workers/LocalWorker.cpp:262
#13 0x00000000004f43f2 in Worker::threadStart (worker=0x7b5920) at source/workers/Worker.cpp:14
#14 0x00002aaaabea0cd4 in ?? () from /usr/lib64/libstdc++.so.6
#15 0x00002aaaac6fd539 in start_thread () from /lib64/libpthread.so.0
#16 0x00002aaaaca0fecf in clone () from /lib64/libc.so.6
```

**Comment by breuner** [comment]
Thanks for reporting this, @glennklockwood . 
I believe this was caused by a buffer overrun in the code path of "--blockvarpct 25" and it should be fixed now with the latest version (v2.1.3). I will leave this issue open until you find the time to confirm.

**Comment by glennklockwood** [comment]
Thank you!  Happy to report that the crash that was happening with c997820 no longer happens in a05cdec when I run

```
./bin/elbencho --timelimit 45 --blockvaralgo fast --blockvarpct 25 --resfile xyz --csvfile xyz.out.csv --threads 16 --files 1 --dirsharing --mkdirs --sync --direct --write --size 1t --block 4k --rand $SCRATCH
```

Problem solved.
